CoreAnalyticSchema/src/au/gov/asd/tac/constellation/graph/schema/analytic/compatibility/AnalyticSchemaV4UpdateProvider.java

```diff
@@ -0,0 +1,85 @@
+import au.gov.asd.tac.constellation.graph.versioning.SchemaUpdateProvider;
+import au.gov.asd.tac.constellation.graph.versioning.UpdateProvider;
+import java.util.HashMap;
+import java.util.Map;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ * Upgrade Vertex Types that have changed.
+ * <p>
+ * The detection regex and validation regex was reviewed and improved.
+ *
+ * @author arcturus
+ */
+@ServiceProvider(service = UpdateProvider.class)
+public class AnalyticSchemaV4UpdateProvider extends SchemaUpdateProvider {
+
+    public static final int SCHEMA_VERSION_THIS_UPDATE = 4;
+
+    @Override
+    protected SchemaFactory getSchema() {
+        return SchemaFactoryUtilities.getSchemaFactory(AnalyticSchemaFactory.ANALYTIC_SCHEMA_ID);
+    }
+
+    @Override
+    public int getFromVersionNumber() {
+        return AnalyticSchemaV3UpdateProvider.SCHEMA_VERSION_THIS_UPDATE;
+    }
+
+    @Override
+    public int getToVersionNumber() {
+        return SCHEMA_VERSION_THIS_UPDATE;
+    }
+
+    @Override
+    protected void schemaUpdate(final StoreGraph graph) {
+        final int typeAttribute = AnalyticConcept.VertexAttribute.TYPE.get(graph);
+
+        final Map<String, SchemaVertexType> typesToUpgrade = new HashMap<>();
+        typesToUpgrade.put(AnalyticConcept.VertexType.MD5.getName(), AnalyticConcept.VertexType.MD5);
+        typesToUpgrade.put(AnalyticConcept.VertexType.SHA1.getName(), AnalyticConcept.VertexType.SHA1);
+        typesToUpgrade.put(AnalyticConcept.VertexType.SHA256.getName(), AnalyticConcept.VertexType.SHA256);
+        typesToUpgrade.put(AnalyticConcept.VertexType.COUNTRY.getName(), AnalyticConcept.VertexType.COUNTRY);
+        typesToUpgrade.put(AnalyticConcept.VertexType.GEOHASH.getName(), AnalyticConcept.VertexType.GEOHASH);
+        typesToUpgrade.put(AnalyticConcept.VertexType.MGRS.getName(), AnalyticConcept.VertexType.MGRS);
+        typesToUpgrade.put(AnalyticConcept.VertexType.MD5.getName(), AnalyticConcept.VertexType.MD5);
+        typesToUpgrade.put(AnalyticConcept.VertexType.IPV4.getName(), AnalyticConcept.VertexType.IPV4);
+        typesToUpgrade.put(AnalyticConcept.VertexType.IPV6.getName(), AnalyticConcept.VertexType.IPV6);
+        typesToUpgrade.put(AnalyticConcept.VertexType.EMAIL_ADDRESS.getName(), AnalyticConcept.VertexType.EMAIL_ADDRESS);
+        typesToUpgrade.put(AnalyticConcept.VertexType.HOST_NAME.getName(), AnalyticConcept.VertexType.HOST_NAME);
+        typesToUpgrade.put(AnalyticConcept.VertexType.URL.getName(), AnalyticConcept.VertexType.URL);
+        typesToUpgrade.put(AnalyticConcept.VertexType.TELEPHONE_IDENTIFIER.getName(), AnalyticConcept.VertexType.TELEPHONE_IDENTIFIER);
+
+        for (int vertex = 0; vertex < graph.getVertexCount(); vertex++) {
+            final int vertexId = graph.getVertex(vertex);
+            final SchemaVertexType oldType = graph.getObjectValue(typeAttribute, vertexId);
+
+            if (typesToUpgrade.containsKey(oldType.getName())) {
```

Comment:
This will probably work just fine, but could be improved by comparing more of the type object than just the name.

Reply:
Perhaps @cygnus-x-1 though remember that this is upgrading from v3 so we should be safe to assume what the type is at this point.